retrievalmarket/server/gsunpaidretrieval.go

```diff
@@ -0,0 +1,475 @@
+			return
+		}
+
+		// Send the other peer a message that the transfer has completed
+		if err := g.dtnet.SendMessage(g.ctx, p, respMsg); err != nil {
+			err := fmt.Errorf("failed to send completion message to requestor %s: %w", p, err)
+			g.failTransfer(state, err)
+			return
+		}
+
+		state.cs.status = datatransfer.Completed
+		g.publishDTEvent(datatransfer.Complete, "", state.cs)
+		// Fire markets blocks completed event
+		state.mkts.Status = retrievalmarket.DealStatusCompleted
+		g.publishMktsEvent(retrievalmarket.ProviderEventComplete, *state.mkts)
+
+		log.Infow("successfully sent completion message to requestor", "peer", p)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterRequestorCancelledListener(listener graphsync.OnRequestorCancelledListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterRequestorCancelledListener(func(p peer.ID, request graphsync.RequestData) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request)
+			return
+		}
+
+		state.cs.status = datatransfer.Cancelled
+		g.publishDTEvent(datatransfer.Cancel, "", state.cs)
+		state.mkts.Status = retrievalmarket.DealStatusCancelled
+		g.publishMktsEvent(retrievalmarket.ProviderEventCancelComplete, *state.mkts)
+
+		g.untrackTransfer(p, state.cs.transferID)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterBlockSentListener(listener graphsync.OnBlockSentListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterBlockSentListener(func(p peer.ID, request graphsync.RequestData, block graphsync.BlockData) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request, block)
+			return
+		}
+
+		// When a data transfer is restarted, the requester sends a list of CIDs
+		// that it already has. Graphsync calls the sent hook for all blocks even
+		// if they are in the list (meaning, they aren't actually sent over the
+		// wire). So here we check if the block was actually sent
+		// over the wire before firing the data sent event.
+		if block.BlockSizeOnWire() == 0 {
+			return
+		}
+
+		// Fire block sent event
+		state.cs.sent += block.BlockSizeOnWire()
+		g.publishDTEvent(datatransfer.DataSent, "", state.cs)
+		state.mkts.TotalSent += block.BlockSizeOnWire()
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterNetworkErrorListener(listener graphsync.OnNetworkErrorListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterNetworkErrorListener(func(p peer.ID, request graphsync.RequestData, err error) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request, err)
+			return
+		}
+
+		// Don't set the channel state to errored, as a network error is not
+		// necessarily fatal
+		// TODO: If the client always sends a restart message after a network
+		// disconnect, then maybe a network error *should* be fatal
```

Comment:
I believe this is cause channel state errored is a terminal state you can't get out of. A network error should ideally put you in a "Waiting for restart" type state. Because a client doesn't ALWAYS send a restart. The network error could be cause they closed their laptop and won't be trying again for a day. This is partly why Jacob had to write the cleanup job. And keeping the request around for anything more than maybe a minute is a DOS vector, especially since we're tracking all state in memory now.

Reply:
Thanks, I guess we can just error out then. If the client sends a restart request it will just open a new transfer (for the remaining data)